hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakeAwareAccountsCommitsInterceptor.java

```diff
@@ -0,0 +1,335 @@
+	private final Supplier<MerkleNetworkContext> networkCtx;
+	private final RewardCalculator rewardCalculator;
+	private final SideEffectsTracker sideEffectsTracker;
+	private final GlobalDynamicProperties dynamicProperties;
+	private final StakePeriodManager stakePeriodManager;
+	private final StakeInfoManager stakeInfoManager;
+
+	private boolean rewardsActivated;
+	private boolean rewardBalanceIncreased;
+	private boolean[] hasBeenRewarded;
+
+	private static final Logger log = LogManager.getLogger(StakeAwareAccountsCommitsInterceptor.class);
+	private static final long STAKING_FUNDING_ACCOUNT_NUMBER = 800L;
+
+	public StakeAwareAccountsCommitsInterceptor(
+			final SideEffectsTracker sideEffectsTracker,
+			final Supplier<MerkleNetworkContext> networkCtx,
+			final GlobalDynamicProperties dynamicProperties,
+			final RewardCalculator rewardCalculator,
+			final StakeChangeManager stakeChangeManager,
+			final StakePeriodManager stakePeriodManager,
+			final StakeInfoManager stakeInfoManager) {
+		super(sideEffectsTracker);
+		this.stakeChangeManager = stakeChangeManager;
+		this.networkCtx = networkCtx;
+		this.rewardCalculator = rewardCalculator;
+		this.sideEffectsTracker = sideEffectsTracker;
+		this.dynamicProperties = dynamicProperties;
+		this.stakePeriodManager = stakePeriodManager;
+		this.stakeInfoManager = stakeInfoManager;
+	}
+
+	@Override
+	public void preview(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		// if the rewards are activated previously they will not be activated again
+		rewardsActivated = rewardsActivated || networkCtx.get().areRewardsActivated();
+		rewardBalanceIncreased = false;
+
+		// Iterate through the change set, maintaining two invariants:
+		//   1. At the beginning of iteration i, any account that is rewardable due to change in balance or
+		//      stakedAccountId or stakedNodeId or declineRewards fields. Also checks the balance of funding account
+		//      0.0.800 if it has reached the ONE TIME threshold to activate staking.
+		//      NOTE that this activation happens only once.
+		//   2. Any account whose stakedToMe balance is affected by a change in the [0, i) range has
+		//      been, if not already present, added to the pendingChanges; and its changes include its
+		//      new STAKED_TO_ME change
+		updateAccountStakes(pendingChanges);
+
+		// Iterate through the change set again to update node stakes; we do this is in a
+		// separate loop to ensure all STAKED_TO_ME fields have their final values
+		updateNodeStakes(pendingChanges);
+		updateFundingRewardBalances(pendingChanges);
+
+		super.preview(pendingChanges);
+
+		checkStakingRewardsActivation(pendingChanges);
+	}
+
+	private void updateAccountStakes(final EntityChangeSet<AccountID, MerkleAccount, AccountProperty> pendingChanges) {
+		// boolean to track if the account has been rewarded already with one of the pending changes
+		hasBeenRewarded = new boolean[64];
```

Comment:
Can eliminate this array creation, and just begin each iteration of the below `for` with,
```
hasBeenRewarded[i] = false;
```